### PR TITLE
feat: rules refactoring / add separate arazzo rule-set

### PR DESCRIPTION
**Pull request by DmitryAnansky**
## What/Why/How?

- Refactoring rules structure.
- Add separate Arazzo ruleset.
- Update docs structure for rules.
- Removed the `type:none` sourceDescriptions Arazzo extension.
- Add additional lint rule for sourceDescriptions:

sourceDescriptions | [Source Description Object] | REQUIRED. A list of source descriptions (such as an OpenAPI description) this Arazzo Description SHALL apply to. The list MUST have at least one entry.
-- | -- | --

TODO: 

- [x] - add docs redirects [#1514](https://github.com/Redocly/marketing-site-portal/pull/1514)

## Reference

https://github.com/Redocly/redocly/issues/11343

## Testing

chore(all): dummy pr to test new cli and core [#12281](https://github.com/Redocly/redocly/pull/12281)

## Check yourself

- [x] Code changed? - Tested with redoc/reference-docs/workflows (internal)
- [x] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/config/types.ts [hunk 1]
```diff
@@ -58,13 +59,13 @@ export type StyleguideRawConfig<T = undefined> = {
...
 
-  rules?: RuleMap<BuiltInCommonOASRuleId, RuleConfig, T>;
+  rules?: RuleMap<BuiltInCommonOASRuleId, RuleConfig, T> | RuleMap<BuiltInCommonArazzoRuleId, RuleConfig, T>;
   oas2Rules?: RuleMap<BuiltInOAS2RuleId, RuleConfig, T>;
   oas3_0Rules?: RuleMap<BuiltInOAS3RuleId, RuleConfig, T>;
   oas3_1Rules?: RuleMap<BuiltInOAS3RuleId, RuleConfig, T>;
   async2Rules?: RuleMap<BuiltInAsync2RuleId, RuleConfig, T>;
   async3Rules?: RuleMap<BuiltInAsync3RuleId, RuleConfig, T>;
-  arazzoRules?: RuleMap<BuiltInArazzoRuleId, RuleConfig, T>;
+  arazzo1Rules?: RuleMap<BuiltInArazzo1RuleId, RuleConfig, T>;
```
Comment (DmitryAnansky) [inline packages/core/src/config/types.ts:68]: There actually might be no Arazzo version 1 specific rules at the moment.

File: packages/core/src/config/minimal.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ const minimal: PluginStyleguideConfig<'built-in'> = {
     'no-enum-type-mismatch': 'warn',
     'paths-kebab-case': 'off',
     spec: 'error',
+    struct: 'error',
```
Comment (DmitryAnansky) [inline packages/core/src/config/minimal.ts:33]: This probably must be `off`.
Or we will get duplicates.

File: packages/core/src/config/recommended.ts [hunk 3]
```diff
@@ -108,23 +109,6 @@ const recommended: PluginStyleguideConfig<'built-in'> = {
     'channels-kebab-case': 'off',
     'no-channel-trailing-slash': 'off',
   },
-  arazzoRules: {
```
Comment (DmitryAnansky) [inline packages/core/src/config/recommended.ts:111]: Keep or remove from other rule-sets?

Comment (DmitryAnansky) [reply packages/core/src/config/recommended.ts:111]: we should keep it.

File: .changeset/purple-lobsters-stare.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@redocly/openapi-core": patch
+"@redocly/cli": patch
+---
+
+Updated sourceDescriptions to require a valid type, to align with the Arazzo specification.
```
Comment (adamaltman) [inline .changeset/purple-lobsters-stare.md:6]:
```suggestion
Updated `sourceDescriptions` to enforce a valid type field, ensuring compliance with the Arazzo specification.
```

File: .changeset/strong-ties-argue.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+"@redocly/openapi-core": patch
+"@redocly/cli": patch
+---
+
+Introduced `struct` rule. Added Arazzo version 1 ruleset.
```
Comment (JLekawa) [inline .changeset/strong-ties-argue.md:6]:
```suggestion
Introduced `struct` rule.
Added Arazzo version 1 ruleset.
```

**Review by JLekawa — APPROVED** [review]

**Review by adamaltman — APPROVED** [review]

**Review by RomanHotsiy — COMMENTED** [review]
make sure lintignore works with fallback from spec to struct

**Comment by DmitryAnansky** [comment]
> lintignore

Are you talking about `--generate-ignore-file` option?
Tested with `spec`:

Screenshot 2024-11-25 at 11 24 06

Screenshot 2024-11-25 at 11 24 12


Tested with `struct`:

Screenshot 2024-11-25 at 11 24 38

Screenshot 2024-11-25 at 11 24 52

Screenshot 2024-11-25 at 11 24 58

Please let me know if you have other test cases in mind.

File: .gitignore [hunk 6]
```diff
@@ -13,3 +13,4 @@ output/
 *.tgz
 redoc-static.html
 packages/cli/README.md
+redocly.yaml
```
Comment (tatomyr) [inline .gitignore:16]: Please remove this.

**Review by tatomyr — COMMENTED** [review]
Left a few comments.

File: packages/core/src/types/redocly-yaml.ts [hunk 7]
```diff
@@ -51,9 +54,68 @@ const builtInCommonOASRules = [
   'no-required-schema-properties-undefined',
 ] as const;
 
+const builtInCommonArazzoRules = [
```
[the remark this answers is not in the document]
Comment (DmitryAnansky) [reply packages/core/src/types/redocly-yaml.ts:58]: Not really - but it should only have `struct`

File: packages/core/src/types/arazzo.ts [hunk 8]
```diff
@@ -50,10 +50,8 @@ const SourceDescriptions: NodeType = {
   items: (value: any) => {
     if (value?.type === 'openapi') {
       return 'OpenAPISourceDescription';
-    } else if (value?.type === 'arazzo') {
-      return 'ArazzoSourceDescription';
     } else {
-      return 'NoneSourceDescription';
+      return 'ArazzoSourceDescription';
```
[the remark this answers is not in the document]
Comment (DmitryAnansky) [reply packages/core/src/types/arazzo.ts:54]: This  has two options:

1. Not throw any error in this case and validate with 'sourceDescription-type': 'error' rule.
In this case we can get good location for the problem, like
```
The `type` property of the `sourceDescription` object must be either `openapi` or `arazzo`.

 8 |
 9 | sourceDescriptions:
10 |   - name: museum-api
11 |     type: openapid
12 |     url: ../openapi.yaml
```

2. Throw an error:
```
    if (value?.type === 'openapi') {
      return 'OpenAPISourceDescription';
    } else if (value?.type === 'arazzo') {
      return 'ArazzoSourceDescription';
    } else {
      throw new Error(`SourceDescription type ${value?.type} is not supported`);
    }
```
Then we can remove the rule `sourceDescription-type`, but output is worse in my opinion, as there no good  way we can help user to show location :

Screenshot 2024-11-28 at 14 50 56

File: packages/core/src/rules/arazzo/sourceDescription-type.ts [hunk 9]
```diff
@@ -4,13 +4,14 @@ import type { UserContext } from '../../walk';
...
-        for (const sourceDescription of SourceDescriptions) {
+      enter(sourceDescriptions, { report, location }: UserContext) {
+        if (!sourceDescriptions.length) return;
+        for (const sourceDescription of sourceDescriptions) {
           if (!['openapi', 'arazzo'].includes(sourceDescription?.type)) {
             report({
               message:
                 'The `type` property of the `sourceDescription` object must be either `openapi` or `arazzo`.',
-              location: location.child([SourceDescriptions.indexOf(sourceDescription)]),
+              location: location.child([sourceDescriptions.indexOf(sourceDescription)]),
```
[the remark this answers is not in the document]
Comment (DmitryAnansky) [reply packages/core/src/rules/arazzo/sourceDescription-type.ts:14]: This  has two options:

1. Not throw any error in this case and validate with 'sourceDescription-type': 'error' rule.
In this case we can get good location for the problem, like
```
The `type` property of the `sourceDescription` object must be either `openapi` or `arazzo`.

 8 |
 9 | sourceDescriptions:
10 |   - name: museum-api
11 |     type: openapid
12 |     url: ../openapi.yaml
```

2. Throw an error:
```
    if (value?.type === 'openapi') {
      return 'OpenAPISourceDescription';
    } else if (value?.type === 'arazzo') {
      return 'ArazzoSourceDescription';
    } else {
      throw new Error(`SourceDescription type ${value?.type} is not supported`);
    }
```
Then we can remove the rule `sourceDescription-type`, but output is worse in my opinion, as there no good  way we can help user to show location :

Screenshot 2024-11-28 at 14 50 56

File: packages/core/src/types/redocly-yaml.ts [hunk 10]
```diff
@@ -109,14 +208,14 @@ export type BuiltInAsync2RuleId = typeof builtInAsync2Rules[number];
 
 export type BuiltInAsync3RuleId = typeof builtInAsync3Rules[number];
 
-const builtInArazzoRules = [
-  'spec',
+const builtInArazzo1Rules = [
```
[the remark this answers is not in the document]
Comment (DmitryAnansky) [reply packages/core/src/types/redocly-yaml.ts:205]: `builtInArazzo1Rules` is not 1-to-1 equal to the `ArazzoRuleset` keys as it does not have common rules inside.

File: packages/core/src/config/builtIn.ts [hunk 11]
```diff
@@ -2,11 +2,12 @@ import recommended from './recommended';
 import recommendedStrict from './recommended-strict';
 import all from './all';
 import minimal from './minimal';
+import specArazzo from './spec-arazzo';
```
[the remark this answers is not in the document]
Comment (tatomyr) [reply packages/core/src/config/builtIn.ts:5]: Clarified -- this is for having a ruleset that covers only what is in the Arazzo specification.

I have another idea. What if we introduce a general `spec` ruleset that follows all supported specifications, not only Arazzo? The tool is smart enough to apply only the related rules depending on the flavor. 
 @RomanHotsiy, @adamaltman what do you think?

[the remark this answers is not in the document]
Comment (adamaltman) [reply packages/core/src/config/builtIn.ts:5]: I'm fine with that too.

[the remark this answers is not in the document]
Comment (RomanHotsiy) [reply packages/core/src/config/builtIn.ts:5]: Makes sense.

[the remark this answers is not in the document]
Comment (tatomyr) [reply packages/core/src/config/builtIn.ts:5]: I suggest we start small in this PR -- with only `struct` rule for specifications other that Arazzo, just to have merged this one sooner, and after that we'll enhance it with other rules.

File: .changeset/strong-ties-argue.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
+---
+"@redocly/openapi-core": patch
+"@redocly/cli": patch
+---
+
+Introduced `struct` rule.
+Added Arazzo version 1 ruleset.
```
Comment (tatomyr) [inline .changeset/strong-ties-argue.md:7]:
```suggestion
Introduced `struct` rule. Added a `spec` ruleset which enforces compliance with the specifications.
```

File: .gitignore [hunk 13]
```diff
@@ -12,4 +12,4 @@ output/
 *.tsbuildinfo
 *.tgz
 redoc-static.html
-packages/cli/README.md
+packages/cli/README.md
```
Comment (tatomyr) [inline .gitignore:15]:
```suggestion
packages/cli/README.md

```

File: packages/core/src/config/builtIn.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (DmitryAnansky) [reply packages/core/src/config/builtIn.ts]: Updated, please take a look when you have time.

File: packages/core/src/config/spec.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
+import type { PluginStyleguideConfig } from './types';
+
+const spec: PluginStyleguideConfig<'built-in'> = {
```
Comment (tatomyr) [inline packages/core/src/config/spec.ts:3]: I believe there should be a page describing this ruleset.

File: docs/rules/minimal.md [hunk 15]
```diff
@@ -8,33 +8,33 @@ The rules contained in the minimal ruleset:
 
 Errors:
 
-- [no-server-trailing-slash](./no-server-trailing-slash.md)
-- [no-server-variables-empty-enum](./no-server-variables-empty-enum.md)
-- [no-unresolved-refs](./no-unresolved-refs.md)
+- [no-server-trailing-slash](./oas/no-server-trailing-slash.md)
+- [no-server-variables-empty-enum](./oas/no-server-variables-empty-enum.md)
+- [no-unresolved-refs](./oas/no-unresolved-refs.md)
 - [spec](./spec.md)
```
Comment (tatomyr) [inline docs/rules/minimal.md:14]:
```suggestion
- [struct](./oas/struct.md)
```

File: packages/core/src/config/spec.ts (hunk 14 above)
Comment (DmitryAnansky) [reply packages/core/src/config/spec.ts:3]: I will add file but I think we will need to populate it later as I am missing rules description documentation.
cc: @JLekawa

**Review by tatomyr — APPROVED** [review]
LGTM. Thanks!

File: .changeset/strong-ties-argue.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
+---
+"@redocly/openapi-core": patch
+"@redocly/cli": patch
```
Comment (DmitryAnansky) [inline .changeset/strong-ties-argue.md:3]: - change to 'minor'

File: docs/rules/arazzo/struct.md [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+Ensures that your API document conforms to the [OpenAPI specification](https://spec.openapis.org/arazzo/latest.html#arazzo-specification).
+
+| Arzzo | Compatibility |
+| ----- | ------------- |
+| 1.0.0 | ✅            |
+
+The default setting for this rule (in the `recommended` and `minimal` configuration) is `error`.
+
+This is an essential rule that should not be turned off except in rare and special cases.
```
Comment (JLekawa) [inline docs/rules/arazzo/struct.md:15]:
```suggestion
This is an essential rule. Do not turned it off, except in rare and special cases.
```

File: docs/rules/async/channels-kebab-case.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+Example of an **incorrect** channel:
+
+```yaml
+channels:
+  ticketSales:
+    address: transactions/ticketSales # channel address value checked by rule
+    messages:
+      ticketSaleTransaction:
+        $ref: '#/components/messages/ticketSaleTransaction'
+```
```
Comment (JLekawa) [inline docs/rules/async/channels-kebab-case.md:45]:
```suggestion
```

File: docs/rules/async/channels-kebab-case.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+### Channel rules for AsyncAPI 2.6
+
+The syntax for how the channels are described changed with the AsyncAPI 3.0 release.
+
+If you still AsyncAPI 2.6, this rule works, and checks the channel address used as the key of the `channels` object.
```
Comment (JLekawa) [inline docs/rules/async/channels-kebab-case.md:62]:
```suggestion
If you use AsyncAPI 2.6, this rule works and checks the channel address used as the key of the `channels` object.
```

File: docs/rules/async/no-channel-trailing-slash.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+Example of an **incorrect** channel:
+
+```yaml
+channels:
+  channel1:
+    address: events/trailing/
+    messages:
+      event1:
+        $ref: '#/components/messages/event1'
+```
```
Comment (JLekawa) [inline docs/rules/async/no-channel-trailing-slash.md:47]:
```suggestion
```

File: .changeset/strong-ties-argue.md [hunk 21]
```diff
@@ -0,0 +1,6 @@
+---
+"@redocly/openapi-core": minor
+"@redocly/cli": minor
+---
+
+Introduced the `struct` rule and deprecated the `spec` rule. Added the `spec` ruleset, which enforces compliance with the specifications.
```
Comment (JLekawa) [inline .changeset/strong-ties-argue.md:6]:
```suggestion
Introduced the `struct` rule and deprecated the `spec` rule.
Added the `spec` ruleset, which enforces compliance with the specifications.
```

File: docs/rules/async/no-channel-trailing-slash.md [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+### Channel rules for AsyncAPI 2.6
+
+The syntax for how the channels are described changed with the AsyncAPI 3.0 release.
+
+This rule also works with AsyncAPI 2.6, and checks the channel address used as the key of the `channels` object.
```
Comment (JLekawa) [inline docs/rules/async/no-channel-trailing-slash.md:64]:
```suggestion
This rule also works with AsyncAPI 2.6 and checks the channel address used as the key of the `channels` object.
```

**Pull request merged**